File: Algorithm/Sort/typescript/quickSort.ts
// 시간복잡도: 최선, 평균 nlogn, 최악 n^2
// 공간복잡도: 1

const swap = (array: number[], i: number, j: number) => {
  const temp = array[j]
  array[j] = array[i]
  array[i] = temp
}

// 호어 방식
function quickSortByHoare(nums: number[]) {
  const quickSort = (array: number[], start = 0, end = array.length - 1) => {
    if (start >= end) return

    const pivot = start
    let left = start + 1
    let right = end

    while (left <= right) {
      while (left <= end && array[left] <= array[pivot]) {
        left++
      }
      while (start < right && array[pivot] <= array[right]) {
        right--
      }
      if (left <= right) {
        swap(array, left, right)
      } else {
        swap(array, right, pivot)
      }
    }

    quickSort(array, start, right - 1)
    quickSort(array, right + 1, end)
  }

  quickSort(nums)

  return nums
}

// 로무토 방식
function quickSortByLomuto(nums: number[]) {
  const quickSort = (array: number[], lo = 0, hi = nums.length - 1) => {
    const partition = (lo: number, hi: number) => {
      const pivot = array[hi]
      let left = lo
      for (let right = lo; right < hi; right++) {
        if (array[right] < pivot) {
          swap(array, left, right)
          left++
        }
      }
      swap(array, left, hi)
      return left
    }

    if (lo < hi) {
      const pivot = partition(lo, hi)
      quickSort(array, lo, pivot - 1)
      quickSort(array, pivot + 1, hi)
    }
  }

  quickSort(nums)

  return nums
}

File: Algorithm/Sort/python/quick_sort.py
# 시간복잡도: 최선, 평균 nlogn, 최악 n^2
# 공간복잡도: 1
def quicksort(nums:list, start:int, end:int)->list:
    if start >= end:
        return nums
    
    pivot = start
    
    left = start + 1
    right = end

    while left <= right:
        while left <= end and nums[left] <= nums[pivot]:
            left += 1
        while start < right and nums[pivot] <= nums[right]:
            right -= 1
        if left > right:
            nums[right], nums[pivot] = nums[pivot], nums[right]
        else:
            nums[left], nums[right] = nums[right], nums[left]

    quicksort(nums, start, right - 1)
    quicksort(nums, right + 1, end)
    return nums